### ISSUE TITLE
Arrays with value semantics

### ISSUE THREAD

**Issue by fabianschuiki** [fabianschuiki/low#29]
Arrays in Low shall be treated as a value rather than a pointer, comparable to how C treats struct variables. This implies that pointers are not equivalent to arrays. As a consequence, the index operator shall dereference a pointer before performing index access. The type syntax is as follows:

```
[8]int32   // array of 8 32-bit integers
*[8]int32  // pointer to an array of 8 32-bit integers
[]int32    // array of an unspecified number of 32-bit integers
```

Note that the syntax for the array of unspecified length collides with the slice syntax. How do we resolve this issue? Arrays have value semantics:

```
a : [8]int32
b : [8]int32
a[0] = 1
b = a // copies the entire array a into b
b[0] = 2
// now a[0] = 1
// now b[0] = 2
```

In contrast to C, the address of the array must be explicitly taken in order to pass it around by reference:

```
func sum(*[]int32 a) int32 { ... }

a : [8]int32
result := sum(&a)
```

With arrays having to be pointers explicitly, the way Low treats pointers should be adjusted a little bit. The member access operator `.` already performs automatic pointer dereferencing. This concept shall be extended to the index access operator `[]`:

```
a : [8]int32
b : *[8]int32

a[0] = 1  // accesses a
b[0] = 2  // dereferences and accesses b
```

This concept could be extended to other operations such as addition and subtraction. This would make memory operations more explicit, and pointers less explicit:

```
// less explicit pointers
a : int32
b := &a
a = 1
c := b + 1  // c = 2 rather than the address of a + sizeof(int32)

// more explicit memory operations
a : int32
b : *void = &a
a = 1
c : *int32 = b + sizeof(int32)
```

**tl;dr**: Having operators implicitly dereference pointers causes
- pointer arithmetic to be replaced by explicit memory calculations and
- pointers to become almost transparent, except for assignments

**Comment by trichner** [comment]
I generally like the idea.

Some ideas how to resolve the slice collision: Maybe `[?]int32` or `[-]int32` ?

**Comment by fabianschuiki** [comment]
Looks good. We could also change the slice syntax to `[:]int32` such that we don't have to introduce another token that can go in between the brackets.

**Comment by trichner** [comment]
Sounds good to me.
